tuplex/codegen/src/IteratorContextProxy.cc

```diff
@@ -0,0 +1,494 @@
+            std::string funcName;
+            auto iteratorName = iteratorInfo->iteratorName;
+
+            if(iteratorName == "zip") {
+                return updateZipIndex(builder, iterator, iteratorInfo);
+            }
+
+            if(iteratorName == "enumerate") {
+                return updateEnumerateIndex(builder, iterator, iteratorInfo);
+            }
+
+            assert(iteratorName == "iter");
+            auto iterablesType = iteratorInfo->argsType;
+            auto argsIteratorInfo = iteratorInfo->argsIteratorInfo;
+            if(iterablesType.isIteratorType()) {
+                // iter() call on an iterator, ignore the outer iter and call again
+                assert(argsIteratorInfo.front());
+                return updateIteratorIndex(builder, iterator, argsIteratorInfo.front());
+            }
+
+            if(iterablesType.isListType()) {
+                funcName = "list_iterator_update";
+            } else if(iterablesType == python::Type::STRING) {
+                funcName = "str_iterator_update";
+            } else if(iterablesType == python::Type::RANGE){
+                funcName = "range_iterator_update";
+            } else if(iterablesType.isTupleType()) {
+                funcName = "tuple_iterator_update";
+            } else {
+                throw std::runtime_error("Iterator struct " + _env->getLLVMTypeName(iteratorContextType) + " does not have the corresponding LLVM UpdateIteratorIndex function");
+            }
+
+            // function type: i1(*struct.iterator)
+            FunctionType *ft = llvm::FunctionType::get(llvm::Type::getInt1Ty(_env->getContext()),
+                                                       {llvm::PointerType::get(iteratorContextType, 0)}, false);
+            auto *nextFunc = _env->getModule()->getOrInsertFunction(funcName, ft).getCallee();
+            auto exhausted = builder.CreateCall(nextFunc, iterator);
+            return exhausted;
+        }
+
+        SerializableValue IteratorContextProxy::getIteratorNextElement(llvm::IRBuilder<> &builder,
+                                                                   const python::Type &yieldType,
+                                                                   llvm::Value *iterator,
+                                                                   IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            llvm::Type *iteratorContextType = iterator->getType()->getPointerElementType();
+            std::string funcName;
+            auto iteratorName = iteratorInfo->iteratorName;
+
+            if(iteratorName == "zip") {
+                return getZipNextElement(builder, yieldType, iterator, iteratorInfo);
+            }
+
+            if(iteratorName == "enumerate") {
+                return getEnumerateNextElement(builder, yieldType, iterator, iteratorInfo);
+            }
+
+            assert(iteratorName == "iter");
```

Comment:
same here, let's throw an exception for unsupported iterators...

Reply:
Added error message.